app/src/sensors.c

```diff
@@ -18,65 +18,144 @@ LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
 
 #if ZMK_KEYMAP_HAS_SENSORS
 
-struct sensors_data_item {
-    uint8_t sensor_number;
+struct sensors_item_cfg {
+    uint8_t sensor_index;
+    const struct zmk_sensor_config *config;
     const struct device *dev;
     struct sensor_trigger trigger;
 };
 
-#define _SENSOR_ITEM(node)                                                                         \
+#define _SENSOR_ITEM(idx, node)                                                                    \
     {                                                                                              \
-        .dev = NULL, .trigger = {.type = SENSOR_TRIG_DELTA, .chan = SENSOR_CHAN_ROTATION }         \
+        .dev = DEVICE_DT_GET_OR_NULL(node),                                                        \
+        .trigger = {.type = SENSOR_TRIG_DATA_READY, .chan = SENSOR_CHAN_ROTATION},                 \
+        .config = &configs[idx]                                                                    \
     }
+#define SENSOR_ITEM(idx, _i) _SENSOR_ITEM(idx, ZMK_KEYMAP_SENSORS_BY_IDX(idx))
 
-#define SENSOR_ITEM(idx, _node)                                                                    \
-    COND_CODE_1(DT_NODE_HAS_STATUS(ZMK_KEYMAP_SENSORS_BY_IDX(idx), okay),                          \
-                (_SENSOR_ITEM(ZMK_KEYMAP_SENSORS_BY_IDX(idx))), ({}))
+#define PLUS_ONE(n) +1
+#define ZMK_KEYMAP_SENSORS_CHILD_COUNT (0 DT_FOREACH_CHILD(ZMK_KEYMAP_SENSORS_NODE, PLUS_ONE))
+static struct zmk_sensor_config configs[] = {
 
-static struct sensors_data_item sensors[] = {LISTIFY(ZMK_KEYMAP_SENSORS_LEN, SENSOR_ITEM, (, ), 0)};
+#if ZMK_KEYMAP_SENSORS_CHILD_COUNT > 0
 
-static void zmk_sensors_trigger_handler(const struct device *dev,
-                                        const struct sensor_trigger *trigger) {
-    int err;
-    const struct sensors_data_item *item = CONTAINER_OF(trigger, struct sensors_data_item, trigger);
+#define SENSOR_CHILD_ITEM(node)                                                                    \
+    {                                                                                              \
+        .triggers_per_rotation = DT_PROP_OR(                                                       \
+            node, triggers_per_rotation, DT_PROP(ZMK_KEYMAP_SENSORS_NODE, triggers_per_rotation))  \
```

Comment:
With this using DT_PROP instead of DT_PROP_OR, does that mean this property is required to exist on either the sensors node or the child node, even once we support sensors that aren't rotational?

Reply:
Correct, although when we do that, my thinking is to make this particular property optional, and add other settings here that might be relevant to those, e.g. acceleration, inverted-X, etc.

Reply:
Assuming I have correctly determine which DT bindings map to this, https://github.com/zmkfirmware/zmk/pull/1039/files#diff-d2a624d29650518030a8f943eeffa937b4a0162e9d8ea333d231116fb0d3f2abR22 says this property isn't required. Could lead to some difficult to parse compile errors.

Reply:
Added a new default fallback value in Kconfig. Thoughts?